### PR TITLE
Adapting to ES 2.0.0

### PR DESCRIPTION
**Pull request by antonha**
ES 2.0.0 is out, and I'd like to contribute to getting this plugin into that release.

I did some refactoring in the ComboAnalyzerClass, but most of the things are either changes required by Lucene 5.x or ES 2.x

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -76,6 +77,19 @@
         </dependency>
 
         <dependency>
+            <groupId>junit</groupId>
```
Comment (antonha) [inline pom.xml:80]: This was needed due to ES checking for 'jar hell' when booting, even in test. Junit brings in hamcrest-core, which conflicts with the usage of hamcrest-all in the project's tests.

File: pom.xml [hunk 2]
```diff
@@ -38,8 +38,9 @@
     </parent>
 
     <properties>
-        <elasticsearch.version>1.0.0.RC1</elasticsearch.version>
-        <lucene.version>4.6.0</lucene.version>
+        <elasticsearch.version>2.0.0</elasticsearch.version>
+        <lucene.version>5.2.1</lucene.version>
+        <mvn.java.version>1.7</mvn.java.version>
```
Comment (antonha) [inline pom.xml:43]: Took the liberty to upgrade to java 7.. it's been out for a while at least

As a side note, the project does not build with java8, due to much stricter javadoc requirements there. Opted to not change them in this PR.

File: src/main/assemblies/plugin.xml [hunk 3]
```diff
@@ -5,6 +5,13 @@
         <format>zip</format>
     </formats>
     <includeBaseDirectory>false</includeBaseDirectory>
+    <files>
```
Comment (antonha) [inline src/main/assemblies/plugin.xml:8]: The Plugin architecture has changed, now one specifies stuff in an external properties file.

File: src/main/java/org/apache/lucene/analysis/ComboAnalyzer.java [hunk 4]
```diff
@@ -168,113 +157,109 @@ public ComboAnalyzer disableDeduplication() {
...
+        public boolean incrementToken() throws IOException {
+            return false;
+        }
+    };
 
-        // Duplication of the original reader, to feed all sub-analyzers
-        if (subAnalyzers.length <= 1) {
+    @Override
+    protected TokenStreamComponents createComponents(String fieldName) {
+        return new CombiningTokenStreamComponents(fieldName);
```
Comment (antonha) [inline src/main/java/org/apache/lucene/analysis/ComboAnalyzer.java:169]: So, this is the largest change in code. In Lucene 5, components are more or less forced to be reusable. 

Instead of having an external TokenStreamComponents class, there is now an internal one. The internal tokenstreams class will generate the tokenStreams when the tokenStreams() method is called, not before that. In my opinion, this makes this file a bit more readable, since the state is all embedded in the subclass class instead. 

The basic reuse/deduplication is still the same, but methods are smaller. Would love feedback.

File: src/main/java/org/apache/lucene/analysis/ComboAnalyzerWrapper.java [hunk 5]
```diff
@@ -72,7 +71,7 @@ protected void init() {
         String[] sub = settings.getAsArray("sub_analyzers");
         ArrayList<Analyzer> subAnalyzers = new ArrayList<Analyzer>();
         if (sub == null) {
-            throw new ElasticsearchIllegalArgumentException("Analyzer ["+name+"] analyzer of type ["+NAME+"], must have a \"sub_analyzers\" list property");
+            throw new IllegalArgumentException("Analyzer ["+name+"] analyzer of type ["+NAME+"], must have a \"sub_analyzers\" list property");
```
Comment (antonha) [inline src/main/java/org/apache/lucene/analysis/ComboAnalyzerWrapper.java:74]: The ES version is no more.

File: src/main/java/org/apache/lucene/analysis/ComboAnalyzer.java [hunk 6]
```diff
@@ -168,113 +157,109 @@ public ComboAnalyzer disableDeduplication() {
...
-                logger.debug("Ignoring {}th analyzer [{}]. Could not get a TokenStream.", ignored, i, subAnalyzers[i]);
-            }
-            // Use caching if asked or if required in case of duplicated analyzers
-            if (cacheTokenStreams || hasDuplicatedAnalyzers && duplicatedAnalyzers.contains(subAnalyzers[i])) {
-                CachingTokenStream cache = new CachingTokenStream(tempTokenStreams_local[i]);
-                try {
-                    tempTokenStreams_local[i].reset();
-                    cache.fillCache();
-                } catch (IOException ignored) {
-                    logger.debug("Got an error when caching TokenStream from the {}th analyzer [{}]", i, subAnalyzers[i]);
```
Comment (antonha) [inline src/main/java/org/apache/lucene/analysis/ComboAnalyzer.java:240]: Rather than making debug outputs, I opted to just throw a runtime exception if something failed. It's quite severe if one looses tokens without getting an error.

File: src/main/java/org/apache/lucene/analysis/ReusableTokenStreamComponents.java [hunk 7]
```diff
@@ -1,68 +0,0 @@
-package org.apache.lucene.analysis;
```
Comment (antonha) [inline src/main/java/org/apache/lucene/analysis/ReusableTokenStreamComponents.java:1]: Folded into the ComboAnalyzer class

File: src/main/java/org/apache/lucene/util/ReusableStringReaderCloner.java [hunk 8]
```diff
@@ -35,16 +34,18 @@
...
+    private static Class<Reader> reusableStringReader;
 
-    private ReusableStringReader original;
+    private Reader original;
     private String originalContent;
 
     static {
         try {
-            internalField = ReusableStringReader.class.getDeclaredField("s");
+            reusableStringReader = (Class<Reader>) ReusableStringReaderCloner.class.getClassLoader().loadClass("org.apache.lucene.analysis.ReusableStringReader");
```
Comment (antonha) [inline src/main/java/org/apache/lucene/util/ReusableStringReaderCloner.java:47]: This is a glorious glorious hack.

In ES 2.0, plugins get a separate class loader. Package equality are dependent on BOTH the package name and the classloader that loaded the package. Thus, the package name of this class is no longer equal to the org.apache.lucene.analysis package when loaded in ES (I also moved it to clarify). Instead, we manually load the actual class.

File: src/main/java/org/elasticsearch/index/analysis/ComboAnalysisBinderProcessor.java [hunk 9]
```diff
@@ -1,30 +0,0 @@
-/*
```
Comment (antonha) [inline src/main/java/org/elasticsearch/index/analysis/ComboAnalysisBinderProcessor.java:1]: This was mostly boilerplate, now we instead use the provider directly in the plugin class.

File: src/main/java/org/apache/lucene/util/StringReaderCloner.java [hunk 10]
```diff
@@ -18,9 +18,7 @@
  */
 
 // Using javax instead of java because of JVM security measures!
-package javax.io;
```
Comment (antonha) [inline src/main/java/org/apache/lucene/util/StringReaderCloner.java:21]: No use to have this in javax.io, the class is not private. Now all ReaderCloners are in the same package.

File: src/main/resources/plugin-descriptor.properties [hunk 11]
```diff
@@ -0,0 +1,34 @@
+### mandatory elements for all plugins:
```
Comment (antonha) [inline src/main/resources/plugin-descriptor.properties:1]: This is the definition of the plugin in 2.0.0. I opted to keep the comments from the example file for context, but could also be removed I guess

File: src/test/java/org/apache/lucene/analysis/TestComboAnalyzer.java [hunk 12]
```diff
@@ -82,9 +82,31 @@ public void testMultipleAnalyzersDeduplication() throws IOException {
     }
 
     @Test
+    public void testCanReuseForDifferentReader() throws IOException {
```
Comment (antonha) [inline src/test/java/org/apache/lucene/analysis/TestComboAnalyzer.java:85]: I got paranoid

File: src/test/java/org/elasticsearch/index/analysis/TestIntegration.java [hunk 13]
```diff
@@ -16,16 +19,23 @@
...
-public class TestIntegration extends ElasticsearchIntegrationTest {
+public class TestIntegration extends ESIntegTestCase {
 
     protected static final String INDEX = "some_index";
     protected static final String TYPE = "some_type";
 
     public static final String ANALYZER = "configured_analyzer";
 
-    protected void assertAnalyzesTo(String analyzer, String input, String[] output, int startOffsets[], int endOffsets[], String types[], int posIncrements[]) {
+    protected Settings nodeSettings(int nodeOrdinal) {
```
Comment (antonha) [inline src/test/java/org/elasticsearch/index/analysis/TestIntegration.java:35]: Tests now has to manually declare which plugins to load. Dunno why.

File: src/test/java/org/elasticsearch/index/analysis/TestIntegration.java [hunk 14]
```diff
@@ -84,7 +92,7 @@ public void testAnalysis() throws IOException {
                 new int[]{ 0,  0,  0,  5,  7,  7, 14, 14},
                 new int[]{ 4,  4, 18,  6, 13, 13, 18, 18},
                 null,
-                new int[]{ 1,  0,  0,  1,  1,  0,  1,  0});
+                new int[]{ 0,  0,  0,  1,  2,  2,  3,  3});
```
Comment (antonha) [inline src/test/java/org/elasticsearch/index/analysis/TestIntegration.java:100]: Changed this to mean the actual positions of the tokens, instead of their increment. This is what we are testing after all. Also, the test broke due to ES analysis API changes.

**Comment by antonha** [comment]
Any feedback would be very welcome. Most likely, there is some edge case that I've missed.

**Comment by adrianocrestani** [comment]
+1

**Comment by imranazad** [comment]
I tried running the following command:

plugin.bat install "file:///elasticsearch-2.0.0/plugins/elasticsearch-analysis-combo-es20.zip"

but I'm still getting an error regarding 'plugin-descriptor.properties' even though I can see it exists.

**Comment by imranazad** [comment]
I have built the plugin and I'm now getting the following error:

ERROR: C:\elasticsearch-2.0.0\plugins\elasticsearch-analysis-combo-2.0.0-SNAPSHOT.zip\plugin-descriptor.properties

**Comment by antonha** [comment]
@imranazad The plugin installation from file works fine for me on osx, running ES 2.0.0:

```
❯ bin/plugin install "file:///Users/anton/Desktop/es-plugin-test/elasticsearch-analysis-combo-2.0.0-SNAPSHOT.zip"
-> Installing from file:/Users/anton/Desktop/es-plugin-test/elasticsearch-analysis-combo-2.0.0-SNAPSHOT.zip...
Plugins directory [/Users/anton/Desktop/es-plugin-test/elasticsearch-2.0.0/plugins] does not exist. Creating...
Trying file:/Users/anton/Desktop/es-plugin-test/elasticsearch-analysis-combo-2.0.0-SNAPSHOT.zip ...
Downloading .DONE
Verifying file:/Users/anton/Desktop/es-plugin-test/elasticsearch-analysis-combo-2.0.0-SNAPSHOT.zip checksums if available ...
NOTE: Unable to verify checksum for downloaded plugin (unable to find .sha1 or .md5 file to verify)
Installed analysis-combo into /Users/anton/Desktop/es-plugin-test/elasticsearch-2.0.0/plugins/analysis-combo
```

If I guess correctly from your commands, it seems like you have placed the zip-file in the plugin directory, and then try to run the install script on that file? I don't believe that's how the plugin script is intended to be used. Could you try placing the built .zip-file somewhere else, and running the plugin script pointing to that file instead?

This could also be due to permission issues, or due to the plugin-script not working correctly on windows (but that would be highly unlikely).

Until further proof, I'm going to assume that this is _not_ an issue with this plugin specifically.

**Comment by antonha** [comment]
@ofavre Are you still a maintainer of this repo? If so, could you leave a comment regarding this PR? If not, could you point me in the right direction?

Any help would be greatly appreciated :)

**Comment by rog** [comment]
I got the same error as @imranazad:

`ERROR: Could not find plugin descriptor 'plugin-descriptor.properties' in plugin zip`

I did a `wget "https://github.com/antonha/elasticsearch-analysis-combo/archive/master.zip"`

and then `sudo bin/plugin install "file:///home/user/master.zip"`

**Comment by antonha** [comment]
@rogr This plugin is what ES calls a "Java Plugin", which means (among other things) that the zip-file to be installed needs to be built to contain .jar files with .class files in it. For this plugin, this file is built using maven.

You were trying to install a .zip archive containing the source code. This will never work. To install the plugin, you need to build the file from source (until it makes in into the main branch). Also, you should not be using the master branch of my fork. The es20 branch contains the necessary changes for ES 2.0.

**Comment by rog** [comment]
Yeah, I realised that I will built the plugin from your es20 branch and give you feedback..

Hope @yakaz or somebody merge your work to the main repo.

Thank you.

**Comment by imranazad** [comment]
@antonha Sorry for the late response, it's been manic recently. I placed the plugin in a different location and it worked! Many thanks, I really needed this plugin desperately.

**Comment by imranazad** [comment]
@antonha When I define a mapping I'm getting the following error: 

Unknown Analyzer type [combo] for [combo]

Any ideas what I'm missing? I'm using the mappings on the read me page.

**Comment by felixbarny** [comment]
@yakaz are you going to merge this? This would be really important to us.

**Comment by felixbarny** [comment]
Seems like ES 2.2 needs another adjustment

```
java.lang.IllegalAccessException: Class org.elasticsearch.common.inject.DefaultConstructionProxyFactory$1 can not access a member of class org.elasticsearch.index.analysis.ComboAnalyzerProvider with modifiers ""
```

**Comment by vicmosin** [comment]
So did you guys fix it? I am still receiving

```
ERROR: Could not find plugin descriptor 'plugin-descriptor.properties' in plugin zip
```

during installation..

P.S. I have ES v2.1.0

**Comment by felixbarny** [comment]
Have you found an alternative to this plugin which works with current Elasticsearch versions? Unfortunately, this project does not seem to be maintained anymore :( I'm stuck on older versions because of it.

**Comment by vicmosin** [comment]
@felixbarny Unfortunately no..

**Comment by paulhyo** [comment]
Has everybody found an alternative to this plugin which works with current Elasticsearch versions????

**Comment by Kijewski** [comment]
No, I'm still using the old ES version because of that. IMO this plugin is essential and should be adopted into the core. :-/

**Comment by waizuwolf** [comment]
I tried to port this plugin to ES 2.2.0 , I was able to build it. I made the request to ElasticSearch and got following error :
`java.lang.NoClassDefFoundError: Could not initialize class org.apache.lucene.util.ReaderCloneFactory`
I have posted all the logs @ [StackOverflow](http://stackoverflow.com/questions/39206536/noclassdeffounderror-in-elasticsearch)
Any idea , why this is happening. It's working fine on ES 2.1.2
